File: packages/contracts-python/src/aegis_contracts/cinematic.py
"""Phase 28 cinematic incident replay contracts."""

from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, model_validator

from aegis_contracts.errors import ContractErrorCode, ContractValidationError
from aegis_contracts.primitives import IncidentId, RunId, Sequence, SimTimestamp
from aegis_contracts.versioning import (
    CAMERA_DIRECTIVE_SCHEMA_VERSION,
    CINEMATIC_BEAT_SCHEMA_VERSION,
    CINEMATIC_CHAPTER_SCHEMA_VERSION,
    CINEMATIC_PLAN_SCHEMA_VERSION,
    CINEMATIC_PLAYBACK_STATE_SCHEMA_VERSION,
    CINEMATIC_PROVENANCE_SCHEMA_VERSION,
    PRESENTATION_HINT_SCHEMA_VERSION,
    assert_supported_schema_version,
)


class CinematicErrorCode(StrEnum):
    CINEMATIC_VALIDATION_FAILED = "CINEMATIC_VALIDATION_FAILED"
    CINEMATIC_INCOMPATIBLE = "CINEMATIC_INCOMPATIBLE"
    CINEMATIC_MISSING_REFERENCE = "CINEMATIC_MISSING_REFERENCE"
    CINEMATIC_HINT_BLOCKED = "CINEMATIC_HINT_BLOCKED"
    CINEMATIC_REPLAY_UNAVAILABLE = "CINEMATIC_REPLAY_UNAVAILABLE"


class CinematicBeatKindV1(StrEnum):
    ESTABLISHING = "establishing"
    INCIDENT_ORIGIN = "incident_origin"
    EVIDENCE_FOCUS = "evidence_focus"
    PATH_TRACE = "path_trace"
    AGENT_INVESTIGATION = "agent_investigation"
    RISK_CHANGE = "risk_change"
    PROPOSAL_FOCUS = "proposal_focus"
    APPROVAL_MOMENT = "approval_moment"
    CONSEQUENCE_REVEAL = "consequence_reveal"
    REPORT_FOCUS = "report_focus"
    OVERVIEW = "overview"


class CameraDirectiveKindV1(StrEnum):
    ESTABLISHING = "establishing"
    ENTITY_FOCUS = "entity_focus"
    PATH_TRACE = "path_trace"
    AGENT_FOCUS = "agent_focus"
    APPROVAL = "approval"
    CONSEQUENCE = "consequence"
    OVERVIEW = "overview"


class CinematicProvenanceSourceV1(StrEnum):
    REPLAY_STATE = "replay_state"
    PRESENTATION_HINT = "presentation_hint"
    DEFAULT_PLANNER = "default_planner"


class PresentationHintKindV1(StrEnum):
    EMPHASIS = "emphasis"
    CHAPTER_ANCHOR = "chapter_anchor"
    CAPTION = "caption"


class CinematicSessionModeV1(StrEnum):
    NORMAL = "normal"
    CINEMATIC = "cinematic"


class CinematicDirectorStatusV1(StrEnum):
    IDLE = "idle"
    PLAYING = "playing"
    PAUSED = "paused"
    FREE_CAMERA = "free_camera"


class CinematicPlaybackSpeedV1(StrEnum):
    HALF = "0.5x"
    ONE = "1x"
    TWO = "2x"
    FOUR = "4x"


class Vec3WireV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    x: float
    y: float
    z: float


class CameraBookmarkWireV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    position: Vec3WireV1
    target: Vec3WireV1
    fov: float = Field(gt=0)


class CinematicProvenanceV1(BaseModel):
    """Links a cinematic beat to authoritative replay cursor/entity provenance."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    run_id: RunId = Field(alias="runId")
    sequence: Sequence
    sim_time: SimTimestamp | None = Field(default=None, alias="simTime")
    entity_ids: list[str] = Field(default_factory=list, alias="entityIds")
    incident_id: IncidentId | None = Field(default=None, alias="incidentId")
    source: CinematicProvenanceSourceV1
    state_digest: str | None = Field(default=None, alias="stateDigest", max_length=128)

    @model_validator(mode="after")
    def validate_schema_version(self) -> CinematicProvenanceV1:
        assert_supported_schema_version("cinematic_provenance", self.schema_version)
        if self.schema_version != CINEMATIC_PROVENANCE_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message="Unsupported CinematicProvenance schema version",
                details={"schemaVersion": self.schema_version},
            )
        return self


class CameraDirectiveV1(BaseModel):
    """Presentation-only camera instruction; never mutates domain state."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: str = Field(min_length=1, max_length=128)
    kind: CameraDirectiveKindV1
    focus_entity_ids: list[str] = Field(default_factory=list, alias="focusEntityIds")
    path_entity_ids: list[str] = Field(default_factory=list, alias="pathEntityIds")
    bookmark: CameraBookmarkWireV1 | None = None
    transition_ms: int = Field(default=800, alias="transitionMs", ge=0, le=30_000)
    reduced_motion_jump: bool = Field(default=True, alias="reducedMotionJump")

    @model_validator(mode="after")
    def validate_schema_version(self) -> CameraDirectiveV1:
        assert_supported_schema_version("camera_directive", self.schema_version)
        if self.schema_version != CAMERA_DIRECTIVE_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message="Unsupported CameraDirective schema version",
                details={"schemaVersion": self.schema_version},
            )
        return self


class CinematicBeatV1(BaseModel):
    """One directed cinematic moment tied to an authoritative replay sequence."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: str = Field(min_length=1, max_length=128)
    chapter_id: str = Field(alias="chapterId", min_length=1, max_length=128)
    kind: CinematicBeatKindV1
    sequence: Sequence
    sim_time: SimTimestamp | None = Field(default=None, alias="simTime")
    caption: str = Field(min_length=1, max_length=512)
    entity_ids: list[str] = Field(default_factory=list, alias="entityIds")
    path_entity_ids: list[str] = Field(default_factory=list, alias="pathEntityIds")
    camera_directive_id: str = Field(alias="cameraDirectiveId", min_length=1, max_length=128)
    provenance: CinematicProvenanceV1
    priority: int = Field(default=100, ge=0, le=1000)
    tie_breaker: int = Field(default=0, alias="tieBreaker", ge=0, le=1_000_000)

    @model_validator(mode="after")
    def validate_schema_version(self) -> CinematicBeatV1:
        assert_supported_schema_version("cinematic_beat", self.schema_version)
        if self.schema_version != CINEMATIC_BEAT_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message="Unsupported CinematicBeat schema version",
                details={"schemaVersion": self.schema_version},
            )
        return self


class CinematicChapterV1(BaseModel):
    """Ordered chapter spanning a sequence range over authoritative replay."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: str = Field(min_length=1, max_length=128)
    title: str = Field(min_length=1, max_length=256)
    summary: str = Field(min_length=1, max_length=512)
    from_sequence: Sequence = Field(alias="fromSequence")
    to_sequence: Sequence = Field(alias="toSequence")
    beat_ids: list[str] = Field(default_factory=list, alias="beatIds")
    order: int = Field(ge=0, le=1000)

    @model_validator(mode="after")
    def validate_schema_and_range(self) -> CinematicChapterV1:
        assert_supported_schema_version("cinematic_chapter", self.schema_version)
        if self.schema_version != CINEMATIC_CHAPTER_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message="Unsupported CinematicChapter schema version",
                details={"schemaVersion": self.schema_version},
            )
        if self.to_sequence < self.from_sequence:
            raise ContractValidationError(
                code=ContractErrorCode.VALIDATION_FAILED,
                message="toSequence must be >= fromSequence",
                details={
                    "fromSequence": self.from_sequence,
                    "toSequence": self.to_sequence,
                },
            )
        return self


class PresentationHintV1(BaseModel):
    """Safe scenario presentation hint. Must not reveal hidden causes."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: str = Field(min_length=1, max_length=128)
    scenario_id: str = Field(alias="scenarioId", min_length=1, max_length=128)
    kind: PresentationHintKindV1
    sequence: Sequence
    entity_ids: list[str] = Field(default_factory=list, alias="entityIds")
    caption: str | None = Field(default=None, max_length=512)
    chapter_id: str | None = Field(default=None, alias="chapterId", max_length=128)
    requires_evidence_ids: list[str] = Field(
        default_factory=list, alias="requiresEvidenceIds"
    )
    reveals_hidden_cause: bool = Field(default=False, alias="revealsHiddenCause")
    hidden_cause_id: str | None = Field(default=None, alias="hiddenCauseId", max_length=128)

    @model_validator(mode="after")
    def validate_schema_and_safety(self) -> PresentationHintV1:
        assert_supported_schema_version("presentation_hint", self.schema_version)
        if self.schema_version != PRESENTATION_HINT_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message="Unsupported PresentationHint schema version",
                details={"schemaVersion": self.schema_version},
            )
        if self.reveals_hidden_cause is not False:
            raise ContractValidationError(
                code=ContractErrorCode.VALIDATION_FAILED,
                message="Presentation hints must not reveal hidden causes",
                details={"revealsHiddenCause": self.reveals_hidden_cause},
            )
        if self.hidden_cause_id is not None:
            raise ContractValidationError(
                code=ContractErrorCode.VALIDATION_FAILED,
                message="hiddenCauseId is forbidden on safe presentation hints",
                details={"hiddenCauseId": self.hidden_cause_id},
            )
        return self


class CinematicPlaybackStateV1(BaseModel):
    """Ephemeral cinematic director session state (presentation only)."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    run_id: RunId = Field(alias="runId")
    mode: CinematicSessionModeV1
    status: CinematicDirectorStatusV1
    chapter_index: int = Field(alias="chapterIndex", ge=0)
    beat_index: int = Field(alias="beatIndex", ge=0)
    speed: CinematicPlaybackSpeedV1
    reduced_motion: bool = Field(default=False, alias="reducedMotion")
    captions_enabled: bool = Field(default=True, alias="captionsEnabled")
    active_beat_id: str | None = Field(default=None, alias="activeBeatId", max_length=128)
    active_chapter_id: str | None = Field(
        default=None, alias="activeChapterId", max_length=128
    )

    @model_validator(mode="after")
    def validate_schema_version(self) -> CinematicPlaybackStateV1:
        assert_supported_schema_version("cinematic_playback_state", self.schema_version)
        if self.schema_version != CINEMATIC_PLAYBACK_STATE_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message="Unsupported CinematicPlaybackState schema version",
                details={"schemaVersion": self.schema_version},
            )
        return self


class CinematicPlanV1(BaseModel):
    """Deterministic cinematic plan derived from authoritative replay state."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    run_id: RunId = Field(alias="runId")
    chapters: list[CinematicChapterV1] = Field(min_length=1)
    beats: list[CinematicBeatV1] = Field(min_length=1)
    camera_directives: list[CameraDirectiveV1] = Field(
        min_length=1, alias="cameraDirectives"
    )
    warnings: list[str] = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_schema_version(self) -> CinematicPlanV1:
        assert_supported_schema_version("cinematic_plan", self.schema_version)
        if self.schema_version != CINEMATIC_PLAN_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message="Unsupported CinematicPlan schema version",
                details={"schemaVersion": self.schema_version},
            )
        return self

File: packages/contracts-ts/src/cinematic.ts
/** Phase 28 cinematic incident replay contracts. */

import { z } from 'zod';

import { incidentIdSchema, runIdSchema, sequenceSchema, simTimestampSchema } from './primitives';
import {
  CAMERA_DIRECTIVE_SCHEMA_VERSION,
  CINEMATIC_BEAT_SCHEMA_VERSION,
  CINEMATIC_CHAPTER_SCHEMA_VERSION,
  CINEMATIC_PLAN_SCHEMA_VERSION,
  CINEMATIC_PLAYBACK_STATE_SCHEMA_VERSION,
  CINEMATIC_PROVENANCE_SCHEMA_VERSION,
  PRESENTATION_HINT_SCHEMA_VERSION,
} from './versioning';

const schemaVersionCheck = (expected: number) =>
  z
    .number()
    .int()
    .min(1)
    .refine((v) => v === expected);

export const cinematicErrorCodeSchema = z.enum([
  'CINEMATIC_VALIDATION_FAILED',
  'CINEMATIC_INCOMPATIBLE',
  'CINEMATIC_MISSING_REFERENCE',
  'CINEMATIC_HINT_BLOCKED',
  'CINEMATIC_REPLAY_UNAVAILABLE',
]);

export const cinematicBeatKindSchema = z.enum([
  'establishing',
  'incident_origin',
  'evidence_focus',
  'path_trace',
  'agent_investigation',
  'risk_change',
  'proposal_focus',
  'approval_moment',
  'consequence_reveal',
  'report_focus',
  'overview',
]);

export const cameraDirectiveKindSchema = z.enum([
  'establishing',
  'entity_focus',
  'path_trace',
  'agent_focus',
  'approval',
  'consequence',
  'overview',
]);

export const cinematicProvenanceSourceSchema = z.enum([
  'replay_state',
  'presentation_hint',
  'default_planner',
]);

export const presentationHintKindSchema = z.enum(['emphasis', 'chapter_anchor', 'caption']);

export const cinematicSessionModeSchema = z.enum(['normal', 'cinematic']);

export const cinematicDirectorStatusSchema = z.enum(['idle', 'playing', 'paused', 'free_camera']);

export const cinematicPlaybackSpeedSchema = z.enum(['0.5x', '1x', '2x', '4x']);

export const cameraBookmarkWireSchema = z
  .object({
    position: z
      .object({
        x: z.number(),
        y: z.number(),
        z: z.number(),
      })
      .strict(),
    target: z
      .object({
        x: z.number(),
        y: z.number(),
        z: z.number(),
      })
      .strict(),
    fov: z.number().positive(),
  })
  .strict();

export const cinematicProvenanceSchema = z
  .object({
    schemaVersion: schemaVersionCheck(CINEMATIC_PROVENANCE_SCHEMA_VERSION),
    runId: runIdSchema,
    sequence: sequenceSchema,
    simTime: simTimestampSchema.nullable().optional(),
    entityIds: z.array(z.string().min(1).max(128)).default([]),
    incidentId: incidentIdSchema.nullable().optional(),
    source: cinematicProvenanceSourceSchema,
    stateDigest: z.string().min(1).max(128).nullable().optional(),
  })
  .strict();

export const cameraDirectiveSchema = z
  .object({
    schemaVersion: schemaVersionCheck(CAMERA_DIRECTIVE_SCHEMA_VERSION),
    id: z.string().min(1).max(128),
    kind: cameraDirectiveKindSchema,
    focusEntityIds: z.array(z.string().min(1).max(128)).default([]),
    pathEntityIds: z.array(z.string().min(1).max(128)).default([]),
    bookmark: cameraBookmarkWireSchema.nullable().optional(),
    transitionMs: z.number().int().min(0).max(30_000).default(800),
    reducedMotionJump: z.boolean().default(true),
  })
  .strict();

export const cinematicBeatSchema = z
  .object({
    schemaVersion: schemaVersionCheck(CINEMATIC_BEAT_SCHEMA_VERSION),
    id: z.string().min(1).max(128),
    chapterId: z.string().min(1).max(128),
    kind: cinematicBeatKindSchema,
    sequence: sequenceSchema,
    simTime: simTimestampSchema.nullable().optional(),
    caption: z.string().min(1).max(512),
    entityIds: z.array(z.string().min(1).max(128)).default([]),
    pathEntityIds: z.array(z.string().min(1).max(128)).default([]),
    cameraDirectiveId: z.string().min(1).max(128),
    provenance: cinematicProvenanceSchema,
    priority: z.number().int().min(0).max(1000).default(100),
    tieBreaker: z.number().int().min(0).max(1_000_000).default(0),
  })
  .strict();

export const cinematicChapterSchema = z
  .object({
    schemaVersion: schemaVersionCheck(CINEMATIC_CHAPTER_SCHEMA_VERSION),
    id: z.string().min(1).max(128),
    title: z.string().min(1).max(256),
    summary: z.string().min(1).max(512),
    fromSequence: sequenceSchema,
    toSequence: sequenceSchema,
    beatIds: z.array(z.string().min(1).max(128)).default([]),
    order: z.number().int().min(0).max(1000),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.toSequence < value.fromSequence) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'toSequence must be >= fromSequence',
      });
    }
  });

export const presentationHintSchema = z
  .object({
    schemaVersion: schemaVersionCheck(PRESENTATION_HINT_SCHEMA_VERSION),
    id: z.string().min(1).max(128),
    scenarioId: z.string().min(1).max(128),
    kind: presentationHintKindSchema,
    sequence: sequenceSchema,
    entityIds: z.array(z.string().min(1).max(128)).default([]),
    caption: z.string().max(512).nullable().optional(),
    chapterId: z.string().min(1).max(128).nullable().optional(),
    requiresEvidenceIds: z.array(z.string().min(1).max(128)).default([]),
    revealsHiddenCause: z.boolean().default(false),
    hiddenCauseId: z.string().min(1).max(128).nullable().optional(),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.revealsHiddenCause) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Presentation hints must not reveal hidden causes',
      });
    }
    if (value.hiddenCauseId != null) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'hiddenCauseId is forbidden on safe presentation hints',
      });
    }
  });

export const cinematicPlaybackStateSchema = z
  .object({
    schemaVersion: schemaVersionCheck(CINEMATIC_PLAYBACK_STATE_SCHEMA_VERSION),
    runId: runIdSchema,
    mode: cinematicSessionModeSchema,
    status: cinematicDirectorStatusSchema,
    chapterIndex: z.number().int().min(0),
    beatIndex: z.number().int().min(0),
    speed: cinematicPlaybackSpeedSchema,
    reducedMotion: z.boolean().default(false),
    captionsEnabled: z.boolean().default(true),
    activeBeatId: z.string().min(1).max(128).nullable().optional(),
    activeChapterId: z.string().min(1).max(128).nullable().optional(),
  })
  .strict();

export const cinematicPlanSchema = z
  .object({
    schemaVersion: schemaVersionCheck(CINEMATIC_PLAN_SCHEMA_VERSION),
    runId: runIdSchema,
    chapters: z.array(cinematicChapterSchema).min(1),
    beats: z.array(cinematicBeatSchema).min(1),
    cameraDirectives: z.array(cameraDirectiveSchema).min(1),
    warnings: z.array(z.string().max(512)).default([]),
  })
  .strict();

export type CinematicErrorCode = z.infer<typeof cinematicErrorCodeSchema>;
export type CinematicBeatKind = z.infer<typeof cinematicBeatKindSchema>;
export type CameraDirectiveKind = z.infer<typeof cameraDirectiveKindSchema>;
export type CinematicProvenanceSource = z.infer<typeof cinematicProvenanceSourceSchema>;
export type PresentationHintKind = z.infer<typeof presentationHintKindSchema>;
export type CinematicSessionMode = z.infer<typeof cinematicSessionModeSchema>;
export type CinematicDirectorStatus = z.infer<typeof cinematicDirectorStatusSchema>;
export type CinematicPlaybackSpeed = z.infer<typeof cinematicPlaybackSpeedSchema>;
export type CameraBookmarkWire = z.infer<typeof cameraBookmarkWireSchema>;
export type CinematicProvenanceV1 = z.infer<typeof cinematicProvenanceSchema>;
export type CameraDirectiveV1 = z.infer<typeof cameraDirectiveSchema>;
export type CinematicBeatV1 = z.infer<typeof cinematicBeatSchema>;
export type CinematicChapterV1 = z.infer<typeof cinematicChapterSchema>;
export type PresentationHintV1 = z.infer<typeof presentationHintSchema>;
export type CinematicPlaybackStateV1 = z.infer<typeof cinematicPlaybackStateSchema>;
export type CinematicPlanV1 = z.infer<typeof cinematicPlanSchema>;
